rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
         }
     }
 
+    /**
+     * This class keeps together diagnostics messages and the versioned AST for
+     * which the diagnostics were computed. This is needed to ensure that
+     * builder diagnostics are calculated for the same AST as
+     * supposedly-corresponding analyzer diagnostics.
+     */
+    private static class VersionedDiagnostics {
+        public final VersionedTree tree;
+        public final List<Diagnostic> messages;
+
+        public VersionedDiagnostics(VersionedTree tree, List<Diagnostic> messages) {
+            this.tree = tree;
+            this.messages = messages;
+        }
+
+        public static boolean setIfNewer(AtomicReference<VersionedDiagnostics> box, VersionedDiagnostics maybeNewer) {
+            var success = false;
+            var retry = true;
+            while (retry) {
+                var old = box.get();
+                if (old == null || old.tree.version < maybeNewer.tree.version) {
+                    success = box.compareAndSet(old, maybeNewer);
+                    retry = !success;
+                } else {
+                    retry = false;
+                }
+            }
+            return success;
+        }
+
+        public static List<Diagnostic> union(VersionedDiagnostics... list) {
+            var messages = new ArrayList<Diagnostic>();
+            for (var versionedDiagnostics : list) {
+                if (versionedDiagnostics != null) {
+                    messages.addAll(versionedDiagnostics.messages);
+                }
+            }
+            return messages;
+        }
+    }
+
     private class FileFact {
         private final ISourceLocation file;
-        private volatile List<Diagnostic> parseMessages = Collections.emptyList();
-        private volatile List<Diagnostic> typeCheckerMessages = Collections.emptyList();
-        private final ParametricSummaryBridge summary;
+
+        // To replace old diagnostics when new diagnostics become available in a
+        // thread-safe way, we need to atomically: (1) check if the version of
+        // the new diagnostics is greater than the version of the old
+        // diagnostics; (2) if so, replace old with new. This is why
+        // `AtomicReference` and `VersionedDiagnostics` are needed here.
+        private final AtomicReference<VersionedDiagnostics> diagnosticsParser = new AtomicReference<>();
```

Comment:
NP: maybe we can rename them? as to me it reads like a parser for diagnostics? maybe `parserDiagnostics` is enough?

Reply:
also, if we put an empty version in there, we know that it can never be null, avoid bugs down the line.